File: py/lotka.py
from cmdstanpy import CmdStanModel
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import time
import arviz as az

model = CmdStanModel(stan_file="stan/lotka.stan")

no_repetitions = 50
runtimes = [0] * no_repetitions
esss = np.zeros((no_repetitions, 5))

print("RUNNING STAN BENCHMARK")
for i in range(no_repetitions):
    print(f'ITERATION {i}')
    start = time.perf_counter()
    sol = model.sample(data="data/lotka.json", adapt_delta=.65, chains=6, show_progress=False, iter_warmup=500)
    runtimes[i] = time.perf_counter() - start
    esss[i,:] = az.ess(sol).to_array() / runtimes[i]
    print(esss[i,:])
    # print(az.summary(sol))
print("RUNTIMES:")
print(runtimes)

fig, ax = plt.subplots(1,1)
fig.suptitle("ESS/s for CmdStanPy sampling from the Lotka-Volterra posterior.")
ax.set(
    xlabel=f"Stan, mean ESS/s:\n{np.round(np.mean(esss, axis=0))}",
    ylabel="ESS/s [1/s]", ylim=[0, 1000],
    xlim=[.5, 5.5], yticks=[0,200,400,600,800,1000]
)
ax.grid(True)
# sns.boxplot(ax=ax, y=runtimes)
sns.boxplot(ax=ax, data=esss)
sns.swarmplot(ax=ax, data=esss, color="red")
plt.tight_layout()
fig.savefig('figs/lotka_py.png')

File: jl/lotka.jl
using DiffEqBayes, DynamicHMC
using Distributions, BenchmarkTools
using OrdinaryDiffEq, RecursiveArrayTools, ParameterizedFunctions
using Plots
using StatsPlots
using JSON

f = @ode_def LotkaVolterraTest begin
    dx = a*x - b*x*y
    dy = -c*y + d*x*y
end a b c d
u0 = [1.0,1.0]
tspan = (0.0,10.0)
p = [1.5,1.0,3.0,1,0]
prob = ODEProblem(f,u0,tspan,p)
t = collect(range(1,stop=10,length=10))
# sig = 0.49
data = zeros(2, length(t))
jdata = convert(Array, JSON.parsefile("../data/lotka.json")["u"])
for i in 1:2, j in 1:length(t)
    data[i,j] = jdata[j][i]
end
priors = [Truncated(Normal(1.5,0.5),0.5,2.5),Truncated(Normal(1.2,0.5),0,2),Truncated(Normal(3.0,0.5),1,4),Truncated(Normal(1.0,0.5),0,2)]

# run this once to compile
turing_inference(prob,Tsit5(),t,data,priors,num_samples=1_000);

no_repetitions = 20
runtimes = zeros(no_repetitions)

println("RUNNING TURING.JL BENCHMARK")
for i in 1:no_repetitions
	println("ITERATION", i)
	runtimes[i] = @elapsed sol = turing_inference(prob,Tsit5(),t,data,priors,num_samples=1_000)
	display(sol)
end
println("RUNTIMES:")
println(runtimes)

pp = plot(
	xlabel="Turing.jl", ylabel="runtime [s]", ylim=[0, maximum(runtimes)],
	title="Wall times for Turing.jl\nsampling from the Lotka-Volterra posterior.",
	size=(600, 500)
)
boxplot!(pp, runtimes)
savefig(pp, "../figs/lotka_jl.png")
